File: leetcode/d/AlienDictionary.java
/**
 * There is a new alien language which uses the latin alphabet. However, the order among letters are unknown to you. You receive a list of words from the dictionary, where words are sorted lexicographically by the rules of this new language. Derive the order of letters in this language.

For example,
Given the following words in dictionary,

[
  "wrt",
  "wrf",
  "er",
  "ett",
  "rftt"
]
The correct order is: "wertf".

Note:
You may assume all letters are in lowercase.
If the order is invalid, return an empty string.
There may be multiple valid order of letters, return any one of them is fine.
* */

public class Solution {
    public String alienOrder(String[] words) {
        Map<Character, HashSet<Character>> adj = new HashMap<Character, HashSet<Character>>();
        HashMap<Character, Boolean> visited = new HashMap();
        HashMap<Character, Boolean> visiting = new HashMap();
        // Error: don't put non existent node in the adjacent list, which makes the node become an zero edge isolated node
        for (String w : words) {
			for (char c : w.toCharArray()) {
				if (!adj.containsKey(c)) {
					adj.put(c, new HashSet<Character>());  
					visited.put(c, false);
					visiting.put(c, false);
				}
			}
		}
		
		// ERROR: must be an O(n^2) operation, compare word with all the words after it, each comparison add an edge		
        for (int i =0; i < words.length - 1; i++) {
            for (int j = i+1; j < words.length; j++){
                addEdge(words[i], words[j], adj);
            }
        }
        
        String rst = topoSort(adj, visited, visiting);
        return rst;
    }
    
    private void addEdge(String w1, String w2, Map<Character, HashSet<Character>> adj) {
        int len = Math.min(w1.length(), w2.length());
        for (int i = 0; i < len; i++) {
            char c1 = w1.charAt(i);
            char c2 = w2.charAt(i);
            if (c1 != c2) {
                adj.get(c1).add(c2);
                return; // Error: must only add the first different char pair as an edge
            }
        }

    }
    
    private String topoSort(Map<Character, HashSet<Character>> adj, HashMap<Character, Boolean> visited, HashMap<Character, Boolean> visiting) {
        StringBuilder postorder = new StringBuilder();

        for (char c : adj.keySet()){ // ERROR: only go through key in the hashmap, don't iterate through chars not occured in the input
            if (!visited.get(c)) {
                if(hasCycle(adj, c, visited, visiting, postorder)) return ""; // if has cycle return empty string
            }
        }
        return postorder.reverse().toString();
    }
    
    // dfs while check cycle
    private boolean hasCycle(Map<Character, HashSet<Character>> adj, char c, HashMap<Character, Boolean> visited, HashMap<Character, Boolean> visiting, StringBuilder postorder) {
        visiting.put(c, true);
        for (char child : adj.get(c)) {
            if (!visited.get(child)) {
                if (visiting.get(child)) return true;
                else {
                    if (hasCycle(adj, child, visited, visiting, postorder)) return true;
                }
            }
        }
        postorder.append(c);
                visiting.put(c, false);

        visited.put(c, true);
        return false;
    }
    
}

File: lc_plusplus/Alien Dictionary.cc
// There is a new alien language which uses the latin alphabet. However, the order among letters are unknown to you. You receive a list of non-empty words from the dictionary, where words are sorted lexicographically by the rules of this new language. Derive the order of letters in this language.
//
// Example 1:
// Given the following words in dictionary,
//
// [
//   "wrt",
//   "wrf",
//   "er",
//   "ett",
//   "rftt"
// ]
// The correct order is: "wertf".
//
// Example 2:
// Given the following words in dictionary,
//
// [
//   "z",
//   "x"
// ]
// The correct order is: "zx".
//
// Example 3:
// Given the following words in dictionary,
//
// [
//   "z",
//   "x",
//   "z"
// ]
// The order is invalid, so return "".
//
// Note:
// You may assume all letters are in lowercase.
// You may assume that if a is a prefix of b, then a must appear before b in the given dictionary.
// If the order is invalid, return an empty string.
// There may be multiple valid order of letters, return any one of them is fine.


class Solution {
public:
    string alienOrder(vector<string>& words) {
        unordered_map<char, unordered_set<char>> adjs;
        unordered_map<char, int> indegrees;
        for (auto& w : words) {
            for (auto c : w) {
                adjs[c];  // collect all vertices
                indegrees[c] = 0;
            }
        }
        for (int i{0}; i < words.size(); ++i) {
            for (int j{i + 1}; j < words.size(); ++j) {
                int k{0};
                for (; k < min(words[i].size(), words[j].size()); ++k) {
                    if (words[i][k] != words[j][k]) {
                        if (adjs[words[i][k]].count(words[j][k]) == 0) {
                            adjs[words[i][k]].insert(words[j][k]);
                            indegrees[words[j][k]]++;
                        }
                        break;
                    }
                }
                if (k == words[j].size() && words[i].size() > k) return ""; // for invalid case ["wrtkj","wrt"]
            }
        }
        string order;
        queue<char> q;
        for (auto& p : indegrees) if(p.second == 0) q.push(p.first);
        while (!q.empty()) {
            auto c = q.front();
            q.pop();
            order += c;
            for (auto ch : adjs[c]) {
                if (indegrees[ch] != 0 && --indegrees[ch] == 0) q.push(ch);
            }
        }
        for (auto& p : indegrees) if (p.second > 0) return ""; // cyclic
        return order;
    }


    string alienOrderDfs(vector<string>& words) {
        unordered_map<char, unordered_set<char>> adjs;
        for (auto& w : words) {
            for (auto c : w) {
                adjs[c];  // collect all vertices
            }
        }
        for (int i{0}; i < words.size(); ++i) {
            for (int j{i + 1}; j < words.size(); ++j) {
                int k{0};
                for (; k < min(words[i].size(), words[j].size()); ++k) {
                    if (words[i][k] != words[j][k]) {
                        adjs[words[i][k]].insert(words[j][k]);
                        break;
                    }
                }
                if (k == words[j].size() && words[i].size() > k) return ""; // for invalid case ["wrtkj","wrt"]
            }
        }
        unordered_set<char> visited;
        unordered_set<char> visiting;
        string postorder;
        for (auto v : adjs) {
            if (!toposort(adjs, v.first, visited, visiting, postorder)) return "";
        }
        reverse(postorder.begin(), postorder.end());
        return postorder;
    }

    bool toposort(unordered_map<char, unordered_set<char>>& adjs, char src, unordered_set<char>& visited, unordered_set<char>& visiting, string& postorder) {
        if (visited.count(src) == 1) return true;
        visiting.insert(src);
        for (auto& sink : adjs[src]) {
            if (visiting.count(sink) == 1 || !toposort(adjs, sink, visited, visiting, postorder)) return false;  // cyclic
        }
        visiting.erase(src);
        visited.insert(src);
        postorder += src;
        return true;
    }
};
